### PR TITLE
DCS-1412 Using same prisoner search strategy for matching

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/arrivals/ArrivalsService.kt [hunk 1]
```diff
@@ -38,13 +37,19 @@ class ArrivalsService(
   }
 
   private fun augmentWithPrisonData(arrival: Arrival): Arrival {
-    val candidates = prisonerSearchService.getCandidateMatches(arrival)
-
-    val matches = candidates.filter { arrival.isMatch(it) }.distinctBy { it.prisonerNumber }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/arrivals/ArrivalsService.kt:43]: no need for custom matching logic

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 2]
```diff
@@ -4,113 +4,88 @@ import org.slf4j.LoggerFactory
...
   }
 
-  fun getCandidateMatches(prisonNumber: String?, pncNumber: String?): List<MatchPrisonerResponse> {
-    val matchesByPrisonNumber = findMatches(prisonNumber, "Prison Number")
-    val matchesByPncNumber = findMatches(pncNumber, "PNC Number")
+  fun findPotentialMatches(request: MatchPrisonersRequest): List<PotentialMatch> {
+    val results =
+      findMatches(request.prisonNumber, "Prison Number") + //
+        findMatches(request.pncNumber, "PNC Number") + //
+        findNameAndDobMatches(request)
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:30]: comments just to beat formatter into submission

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt [hunk 3]
```diff
@@ -17,9 +17,12 @@ data class MatchPrisonersRequest(
...
   val lastName: String,
 
-  @Schema(description = "Date of birth", example = "1971-02-01")
+  @Schema(
+    description = "Date of birth (If absent then, will not perform name and date of birth search)",
+    example = "1971-02-01"
+  )
   @field:NotNull
-  val dateOfBirth: LocalDate,
+  val dateOfBirth: LocalDate?,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt:25]: we now use this for performing matches and its possible (but unlikely) for BASM moves to be missing dobs

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/response/MatchPrisonerResponse.kt [hunk 4]
```diff
@@ -14,7 +14,7 @@ data class MatchPrisonerResponse(
 
   @JsonFormat(pattern = "yyyy-MM-dd")
   val dateOfBirth: LocalDate,
-  val prisonerNumber: String?,
+  val prisonerNumber: String,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/response/MatchPrisonerResponse.kt:17]: prisoner number is always going to be present for existing prisoners

File: src/test/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/arrivals/ArrivalsServiceTest.kt [hunk 5]
```diff
@@ -161,57 +139,6 @@ class ArrivalsServiceTest {
     }
   }
 
-  @Nested
-  inner class TestMovementMatching {
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/arrivals/ArrivalsServiceTest.kt:165]: no longer needed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchApiClient.kt [hunk 6]
```diff
@@ -27,10 +29,10 @@ class PrisonerSearchApiClient(@Qualifier("prisonerSearchApiWebClient") private v
       ?.firstOrNull()
   }
 
-  fun matchPncNumbersByPrisonerNumbers(matchByPrisonerNumberRequest: MatchByPrisonerNumberRequest): List<PrisonerAndPncNumber> {
+  fun matchPncNumbersByPrisonerNumbers(prisonerNumbers: List<String>): List<PrisonerAndPncNumber> {
     return webClient.post()
       .uri("/prisoner-search/prisoner-numbers")
-      .bodyValue(matchByPrisonerNumberRequest)
+      .bodyValue(MatchByPrisonerNumberRequest(prisonerNumbers))
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchApiClient.kt:35]: pushed this down to neaten up calling code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 7]
```diff
@@ -4,113 +4,88 @@ import org.slf4j.LoggerFactory
...
+  private fun findNameAndDobMatches(request: MatchPrisonersRequest): List<PotentialMatch> {
+    return if (request.dateOfBirth == null) emptyList()
+    else
+      client.matchPrisonerByNameAndDateOfBirth(
+        SearchByNameAndDateOfBirth(
+          request.firstName, request.lastName, request.dateOfBirth
+        )
+      ).map {
+        it.toPotentialMatch()
+      }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:52]: could rewrite to a `?.let` but it was a bit less easy to read

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 8]
```diff
@@ -4,113 +4,88 @@ import org.slf4j.LoggerFactory
...
 import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.request.MatchPrisonerRequest
 import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.request.MatchPrisonersRequest
 import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.response.MatchPrisonerResponse
-import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.response.PrisonerAndPncNumber
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.response.Prisoner
 
 @Service
 class PrisonerSearchService(@Autowired private val client: PrisonerSearchApiClient) {
 
-  fun getCandidateMatches(arrival: Arrival): List<MatchPrisonerResponse> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:20]: this class no longer exposes/returns MatchPrisonerResponses - only prisoner details and potential matches

**Review by GurnankCheema — APPROVED** [review]
LGTM

**Pull request merged**